Mathlib/RingTheory/Polynomial/DegreeLT.lean

```diff
@@ -0,0 +1,218 @@
+
+lemma addLinearEquiv_apply' {R : Type*} [Ring R] (f) :
+    ((addLinearEquiv R m n f).1 : R[X]) = f %ₘ (X ^ m) ∧
+      ((addLinearEquiv R m n f).2 : R[X]) = f /ₘ (X ^ m) := by
+  rw [and_comm, eq_comm, eq_comm (b := _ %ₘ _)]
+  nontriviality R; refine div_modByMonic_unique _ _ (monic_X_pow _) ⟨?_, ?_⟩
+  · rw [← addLinearEquiv_symm_apply', LinearEquiv.symm_apply_apply]
+  · rw [degree_X_pow, ← mem_degreeLT]; exact Subtype.prop _
+
+lemma addLinearEquiv_apply_fst {R : Type*} [Ring R] (f) :
+    ((addLinearEquiv R m n f).1 : R[X]) = f %ₘ (X ^ m) :=
+  (addLinearEquiv_apply' f).1
+
+lemma addLinearEquiv_apply_snd {R : Type*} [Ring R] (f) :
+    ((addLinearEquiv R m n f).2 : R[X]) = f /ₘ (X ^ m) :=
+  (addLinearEquiv_apply' f).2
+
+lemma addLinearEquiv_apply {R : Type*} [Ring R] (f) :
+    addLinearEquiv R m n f =
+      (⟨f %ₘ (X ^ m), addLinearEquiv_apply_fst f ▸ Subtype.prop _⟩,
+      ⟨f /ₘ (X ^ m), addLinearEquiv_apply_snd f ▸ Subtype.prop _⟩) :=
+  Prod.ext (Subtype.ext <| addLinearEquiv_apply_fst f) (Subtype.ext <| addLinearEquiv_apply_snd f)
+
+end degreeLT
+
+section taylor
+
+variable {R : Type*} [CommRing R] {r : R} {m n : ℕ} {s : R} {f g : R[X]}
+
+/-- The map `taylor r` induces an automorphism of the module `R[X]_n` of polynomials of
+degree `< n`. -/
+noncomputable def taylorLinear (r : R) (n) : (R[X]_n) ≃ₗ[R] (R[X]_n) where
+  toFun     := fun P ↦ ⟨P.1.taylor r, mem_degreeLT.2 <| (degree_taylor ..).trans_lt <|
+    mem_degreeLT.1 P.2⟩
+  invFun    := fun P ↦ ⟨P.1.taylor (-r), mem_degreeLT.2 <| (degree_taylor ..).trans_lt <|
+    mem_degreeLT.1 P.2⟩
+  map_add'  := fun _ _ ↦ Subtype.ext <| map_add ..
+  map_smul' := fun _ _ ↦ Subtype.ext <| map_smul ..
+  left_inv  := fun _ ↦ Subtype.ext <| (taylorEquiv r).symm_apply_apply _
+  right_inv := fun _ ↦ Subtype.ext <| (taylorEquiv r).apply_symm_apply _
+
+@[simp] lemma taylorLinear_apply (P : R[X]_n) :
+    (taylorLinear r n P : R[X]) = (P : R[X]).taylor r :=
+  rfl
+
+@[simp] theorem taylor_det' :
+    LinearMap.det (taylorLinear r n : (R[X]_n) →ₗ[R] _) = 1 := by
+  nontriviality R
+  rw [← LinearMap.det_toMatrix (degreeLT.basis R n),
+    Matrix.det_of_upperTriangular, Fintype.prod_eq_one]
+  · intro i
+    rw [LinearMap.toMatrix_apply, degreeLT.basis_repr, ← natDegree_X_pow (R:=R) (i:ℕ),
+      LinearEquiv.coe_coe, taylorLinear_apply, degreeLT.basis_val, taylor_coeff_natDegree,
+      leadingCoeff_X_pow]
+  · intro i j hji
+    rw [LinearMap.toMatrix_apply, LinearEquiv.coe_coe, degreeLT.basis_repr, taylorLinear_apply,
+      degreeLT.basis_val, coeff_eq_zero_of_degree_lt]
+    · rwa [degree_taylor, degree_X_pow, Nat.cast_lt, Fin.val_fin_lt]
+
+@[simp] theorem taylor_det :
+    (taylorLinear r n).det = 1 :=
+  Units.ext <| by rw [LinearEquiv.coe_det, taylor_det', Units.val_one]
+
+/-- Gluing of the `taylorLinear` automorphisms. -/
+noncomputable def taylorProd (r : R) (m n) : ((R[X]_m) × (R[X]_n)) ≃ₗ[R] ((R[X]_m) × (R[X]_n)) :=
+  (taylorLinear r m).prodCongr (taylorLinear r n)
```

Comment:
Does this really need to be its own definition?

Reply:
it is used in the resultant branch. in general having more definitions shortens proofs.